bsp/Src/I2C.c

```diff
@@ -0,0 +1,423 @@
+/*
+ * BSP_I2C.c
+ *
+ *  Created on: Nov 2, 2024
+ *      Author: shaun
+ */
+#include "I2C.h"
+#include "stm32xx_hal.h"
+#include "FreeRTOS.h"
+#include "queue.h"
+#include "semphr.h"
+#include "stdbool.h"
+#include "task.h"
+#include "assert.h"
+
+/* Used Functions */
+void Single_I2C_Init(I2C_HandleTypeDef*);
+
+
+typedef struct {
+	I2C_HandleTypeDef *hi2c;
+	uint8_t deviceAddr;
+	uint8_t pDataBuff;
+	uint8_t payloadsLeft;
+} DataInfo_t;
+
+// Define queue sizes for each I2C peripheral if not done so already
+#ifdef I2C1
+#ifndef I2C1_QUEUE_SIZE
+#define I2C1_QUEUE_SIZE 128
+#endif
+
+#ifndef I2C1_MAX_PAYLOADS
+#define I2C1_MAX_PAYLOADS 16
+#endif
+
+static DataInfo_t I2C1_DataStore[I2C1_QUEUE_SIZE];
+static QueueHandle_t I2C1_Queue;
+static StaticQueue_t I2C1_DataQueue;
+I2C_HandleTypeDef *last_hi2c1;
+#endif
+
+#ifdef I2C2
+#ifndef I2C2_QUEUE_SIZE 
+#define I2C2_QUEUE_SIZE 128
+#endif
+
+#ifndef I2C2_MAX_PAYLOADS
+#define I2C2_MAX_PAYLOADS 16
+#endif
+
+static DataInfo_t I2C2_DataStore[I2C2_QUEUE_SIZE];
+static StaticQueue_t I2C2_DataQueue;
+static QueueHandle_t I2C2_Queue;
+I2C_HandleTypeDef *last_hi2c2;
+#endif
+
+#ifdef I2C3
+#ifndef I2C3_QUEUE_SIZE 
+#define I2C3_QUEUE_SIZE 128
+#endif
+
+#ifndef I2C3_MAX_PAYLOADS
+#define I2C3_MAX_PAYLOADS 16
+#endif
+
+static DataInfo_t I2C3_DataStore[I2C3_QUEUE_SIZE];

```

Comment:
This doesn't need to be of type `DataInfo_t`. You can just have it be `[queue_size * item_length]` instead and have it of type `uint8_t[]`. You also won't have to cast it to a `uint8_t*` later on when calling `QueueCreate`.

Reply:
Understandable. I do believe that keeping the same standard with BSP CAN, and the fact that there do exist chips without i2c (albeit rare), having this already implemented would cause no issue. Do tell if you disagree with this ideology.

Reply:
Fair enough.